datafusion/src/datasource/object_store/mod.rs

```diff
@@ -0,0 +1,130 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Object Store abstracts access to an underlying file/object storage.
+
+pub mod local;
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::pin::Pin;
+use std::sync::{Arc, RwLock};
+
+use async_trait::async_trait;
+use futures::{AsyncRead, Stream};
+
+use local::LocalFileSystem;
+
+use crate::error::{DataFusionError, Result};
+use chrono::Utc;
+
+/// Object Reader for one file in a object store
+#[async_trait]
+pub trait ObjectReader {
+    /// Get reader for a part [start, start + length] in the file asynchronously
+    async fn chunk_reader(&self, start: u64, length: usize)
+        -> Result<Arc<dyn AsyncRead>>;
+
+    /// Get length for the file
+    fn length(&self) -> u64;
+}
+
+/// File meta we got from object store
+#[derive(Debug)]
+pub struct FileMeta {
+    /// Path of the file
+    pub path: String,
+    /// Last time the file was modified in UTC
+    pub last_modified: Option<chrono::DateTime<Utc>>,
+    /// File size in total
+    pub size: u64,
+}
+
+/// Stream of files get listed from object store
+pub type FileMetaStream =
+    Pin<Box<dyn Stream<Item = Result<FileMeta>> + Send + Sync + 'static>>;
+
+/// A ObjectStore abstracts access to an underlying file/object storage.
+/// It maps strings (e.g. URLs, filesystem paths, etc) to sources of bytes
+#[async_trait]
+pub trait ObjectStore: Sync + Send + Debug {
+    /// Returns all the files in path `prefix` asynchronously.
+    async fn list(&self, prefix: &str) -> Result<FileMetaStream>;
+
+    /// Get object reader for one file
+    fn file_reader(&self, file: FileMeta) -> Result<Arc<dyn ObjectReader>>;
```

Comment:
The new `file_reader` and `chunk_reader` method renames makes the code much easier to follow :+1: 

I noticed `file_reader` is still defined as sync method here, curious what's your thoughts on my comment at https://github.com/apache/arrow-datafusion/pull/950#discussion_r697552520?

Reply:
Sorry I didn't fully understand your concern at the first time, fixed now.

Reply:
👍 for the method renames also 😉 

> Maybe we should keep ObjectStore::get_reader as async? Otherwise, it looks like every call to ObjectReader::get_reader would need to open a new file descriptor.

- `ObjectStore` can cache the file descriptor if it wants to, this just forces it to get it lazily
- Re-using the file descriptor means that the various AsyncRead objects generated might interfer with each other if used in parallel, no? (calling `read` on one might move the cursor thus moving the cursor of the other one`). Is this ok?

Reply:
That's a good point @rdettai , we could keep it as a sync method to force lazy initialization. I also agree that sharing a single file descriptor with subsequent read_chunk calls would limit the concurrency to 1. This goes to my second comment in https://github.com/apache/arrow-datafusion/pull/950#discussion_r697552520: making self mut for `read_chunk` method, so we can implement these caching lazily.

Reply:
Sorry for going back and forth on this one. But I think both you and @rdettai were correct that it would be better to keep `file_reader` sync so we can force io into `chunk_reader` lazily. So I think it's best if we could revert the async change for `file_reader`.

Reply:
Reverted to sync method now.